### PR TITLE
ipsec: Fix xfrm output mask

### PR DESCRIPTION
**Pull request by pchaigno**
First commit temporarily uses our netlink fork to be able to set the xfrm output mask. Second commit sets said mask in our xfrm rules. See commit messages for details.

I tested by applying the following diff to our existing test and manually checking the source identity is preserved when making requests between nodes:
```diff
diff --git a/test/k8sT/DatapathConfiguration.go b/test/k8sT/DatapathConfiguration.go
index c11628f67..a91bc17bb 100644
--- a/test/k8sT/DatapathConfiguration.go
+++ b/test/k8sT/DatapathConfiguration.go
@@ -214,7 +214,8 @@ var _ = Describe("K8sDatapathConfig", func() {
 
                        deploymentManager.Deploy(helpers.CiliumNamespace, IPSecSecret)
                        deploymentManager.DeployCilium(map[string]string{
-                               "encryption.enabled": "true",
+                               "encryption.enabled":     "true",
+                               "bpf.monitorAggregation": "none",
                        }, DeployCiliumOptionsAndDNS)
                        validateBPFTunnelMap()
                        Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test with IPsec between nodes failed")
@@ -736,6 +737,7 @@ func testPodConnectivityAndReturnIP(kubectl *helpers.Kubectl, requireMultiNode b
        if !res.WasSuccessful() {
                return false, targetIP
        }
+       helpers.HoldEnvironment("Let's check the source identity!")
 
        // HTTP connectivity test
        res = kubectl.ExecPodCmd(randomNamespace, srcPod,
```

```release-note
Fix missing packet mark mask that can cause policy deny drops in IPSec configuration. 
```

### CODE REVIEW

**Comment by pchaigno** [comment]
test-me-please

**Comment by aditighag** [comment]
Are there any valid cases where we would be sending over zero identities over tunnel header? In other words, can we distinguish between a set/unset identity?

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (jrfastab) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: This comment is incorrect see below, leaving for history so I remember what I was thinking.
...
In ipSecReplaceStateOut I believe we want linux_defaults.IPsecMarkMask not the In version of it.

On out we have skb mark like this, 'skb->mark = 0xIDIDFFID' where the 0xID is the 3 bytes of security context split across the mark. And the 0XFF is the encrypt key used to match with IPsec SPI value. So double check please with a test, but I believe its best to drop the "In" above.

Comment (jrfastab) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: Above is wrong (I left though to remind myself I already consider it).

Actually we are just setting the RouteMarkEncrypt=0x0E00 and RouteMarkDecrypt=0x0D00 values. So IPsecMarkMaskIn=0x0F00 is correct, but change to RouteMarkMask so that it reads better and doesn't confuse us later. This is a bit subtle so best to be clear if possible

**Comment by jrfastab** [comment]
Other than small nit about using the default value RouteMarkMask LGTM

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
> Are there any valid cases where we would be sending over zero identities over tunnel header? In other words, can we distinguish between a set/unset identity?

I can't think of any. If we fail to resolve an identity we should fallback to `WORLD_ID`.

Are you thinking we could fallback to an ipcache lookup at the destination node if the identity is unset? That might work, but (1) I'm not sure we always have all the identities on the remote ipcaches and (2) we would need to consider the complexity impact.

**Comment by pchaigno** [comment]
retest-runtime

**Comment by aditighag** [comment]
> > Are there any valid cases where we would be sending over zero identities over tunnel header? In other words, can we distinguish between a set/unset identity?
> 
> I can't think of any. If we fail to resolve an identity we should fallback to `WORLD_ID`.
> 
> Are you thinking we could fallback to an ipcache lookup at the destination node if the identity is unset? That might work, but (1) I'm not sure we always have all the identities on the remote ipcaches and (2) we would need to consider the complexity impact.

I was thinking more from detecting such error conditions. Perhaps we can log a warning?

**Comment by pchaigno** [comment]
> I was thinking more from detecting such error conditions. Perhaps we can log a warning?

Could be a topic for the next SIG Datapath meeting? It would definitely help uncovering such bugs in CI. It may be a bit hard to implement though because we either need to trace traffic from time to time and know which identity should be non-zero, or we need to extend the datapath to report such cases. Maybe there's a middle ground where we point those out in the `cilium monitor` output?

**Comment by jrfastab** [comment]
An observation,

        RouteMarkMask = 0xF00

and
        // IPsecMarkMask is the mask required for the IPsec SPI and encrypt/decrypt bits
        IPsecMarkMask = 0xFF00

So using IPsecMarkMask is slightly different from original patch. LGTM though.

**Review by jrfastab — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
retest-gke

**Review by tklauser — APPROVED** [review]
Vendoring LGTM, thanks!

I assume we will switch back to use `https://github.com/vishvananda/netlink` once vishvananda/netlink#602 is merged?

**Comment by tklauser** [comment]
Runtime tests failed on known flake #14125 (https://jenkins.cilium.io/job/Cilium-PR-Runtime-4.9/3025/), marking as ready to merge.

**Comment by pchaigno** [comment]
> I assume we will switch back to use `https://github.com/vishvananda/netlink` once [vishvananda/netlink#602](https://github.com/vishvananda/netlink/pull/602) is merged?

Yep, exactly. I will probably ping the maintainer after the holidays if not merged by then.

**Pull request merged**
